zc_events/client.py

```diff
@@ -116,49 +118,54 @@ def wait_for_response(self, response_key):
+    def handle_request_event(self, event, view=None, viewset=None, relationship_viewset=None):
         """
         Method to handle routing request event to appropriate view by constructing
         a request object based on the parameters of the event.
         """
-        request = create_django_request_object(event)
-
-        # Call the viewset passing the appropriate params
-        if event.get('id') and event.get('relationship'):
-            result = relationship_viewset.as_view()(request, pk=event.get('id'),
-                                                    related_field=event.get('relationship'))
-        elif request.method == 'GET' and event.get('id') and event.get('related_resource'):
-            result = viewset.as_view({'get': event.get('related_resource')})(request, pk=event.get('id'))
-        elif request.method == 'GET' and event.get('id'):
-            result = viewset.as_view({'get': 'retrieve'})(request, pk=event.get('id'))
-        elif request.method == 'PUT' and event.get('id'):
-            result = viewset.as_view({'put': 'update'})(request, pk=event.get('id'))
-        elif request.method == 'PATCH' and event.get('id'):
-            result = viewset.as_view({'patch': 'partial_update'})(request, pk=event.get('id'))
-        elif request.method == 'DELETE' and event.get('id'):
-            result = viewset.as_view({'delete': 'destroy'})(request, pk=event.get('id'))
-        elif request.method == 'GET':
-            result = viewset.as_view({'get': 'list'})(request)
-        elif request.method == 'POST':
-            result = viewset.as_view({'post': 'create'})(request)
-        elif request.method == 'OPTIONS' and event.get('id'):
-            result = viewset.as_view({
-                'get': 'retrieve',
-                'put': 'update',
-                'patch': 'partial_update',
-                'delete': 'destroy'
-            })(request, pk=event.get('id'))
-        elif request.method == 'OPTIONS':
-            result = viewset.as_view({
-                'get': 'list',
-                'post': 'create',
-            })(request)
+        kwargs_copy = copy.deepcopy(event)
+        request, kwargs = create_django_request_object(kwargs_copy)
+
+        if not any([view, viewset, relationship_viewset]):
+            raise ImproperlyConfigured('handle_request_event must be passed either a view or viewset')
+
+        response_key = kwargs.pop('response_key')
+        pk = kwargs.get('pk', None)
+        relationship = kwargs.get('relationship', None)
+        related_resource = kwargs.pop('related_resource', None)
+
+        if view:
+            handler = view.as_view()
+        elif pk and relationship:
+            handler = relationship_viewset.as_view()
+        elif request.method == 'GET' and pk and related_resource:
+            handler = viewset.as_view({'get': related_resource})
         else:
-            raise MethodNotAllowed(request.method)
+            if pk:
```

Comment:
I would probably move this part into a different private method, and call it `_get_handler_for_viewset`